pkg/types/chain_reader.go

```diff
@@ -0,0 +1,74 @@
+package types
+
+import (
+	"context"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	ErrInvalidType              = InvalidArgumentError("invalid type")
+	ErrInvalidConfig            = InvalidArgumentError("invalid configuration")
+	ErrChainReaderConfigMissing = UnimplementedError("ChainReader entry missing from RelayConfig")
+)
+
+func UnwrapClientError(err error) error {
+	if s, ok := status.FromError(err); ok {
+		switch s.Code() {
+		case codes.Unimplemented:
+			return UnimplementedError(s.Message())
+		case codes.InvalidArgument:
+			return InvalidArgumentError(s.Message())
```

Comment:
I think this is still too presumptuous since GRPC can return `Unimplemented` for its own reasons, and if we convert it then anything looking for that error won't find it (and would have to understand our custom logic). I think we should try an `Is` method instead. It inverts the relationship, so that only the places using the custom errors explicitly opt-in to the fuzzy matching, and removes the needs for silent lossy conversions.

> An error type might provide an Is method so it can be treated as equivalent to an existing error. For example, if MyError defines
	```go
	func (m MyError) Is(target error) bool { return target == fs.ErrExist }
	```
> then Is(MyError{}, fs.ErrExist) returns true. See [syscall.Errno.Is](https://pkg.go.dev/syscall#Errno.Is) for an example in the standard library. An Is method should only shallowly compare err and the target and not call Unwrap on either.

https://pkg.go.dev/errors#Is

Reply:
I think a major reason why @nolag added the Unwrap function originally was so that the errors coming back from a call to a provider method would look the same on the client side regardless of whether LOOP is enabled.  My most recent implementation is the closet we've come so far to realizing that.

But maybe this is something we should just give up on, as it's not possible to ensure they'll be exactly the same no matter what we do?  (eg, wrapped errors will always get flattened into a single string if they pass through LOOP, and there's no way to reconstruct them reliably)?

Without an Unwrap function, they would look completely different when LOOP is enabled vs when it's not enabled.  For example, with LOOP I think the errors returned would always be some sort of gRPC type (maybe even all the same type?) instead of our type.

Is your idea that we would implement `Is()` on the gRPC error types, and look for those on the client side? Or we implement them on our own types? I'm not sure I see a good way in which it could work.

Reply:
The idiomatic way to handle this situation is to have our custom errors implement a more flexible version of `Is` that matches against GRPC errors which were converted/wrapped from the custom errors. This preserves the behavior in a compatible way: in memory raw errors match normally, and converted grpc errors still match when compared with `errors.Is(err, OurCustomErrType{})`. We just have to define:
```go
func (e OurCustomErrorType) Is(err error) bool { ... }
```

Reply:
We could start with pure string matching, but we might want to using some contains logic that would match on a chained error too, with the standard colon separators.

Reply:
I've never needed to do custom errors before, thanks for the better way.  I can do the implementation tomorrow.  I'll look up the standard colon separator to include it as well, it doesn't sound hard so I don't see a reason not to.

Reply:
Ok, I see

I think nearly all of our error strings will have at least one : in them, because just returning a general error like "invalid config" or "invalid type" isn't very useful. Both of those are the same overall error type of `InvalidArgument`, which we can separate out and represent with the gRPC code.  But they need to include some information about which type or which part of the config is having an issue.  Ideally, a json path to the field in a struct which isn't a valid type, or a json path to a configuration param that is invalid, or a description of which method in the ChainReader config doesn't have the right number of arguments to match the contract abi.

For now we could just match the prefix up until the first : and consider the rest to be "details"... that might not be too difficult.  I guess in the long run, what would be ideal is if we override `FromError()` to take the entire tree (or just list?) of errors and convert it into something that can go inside a `status.Status` object and get serialized.  Then deserialize it with `Is()` for comparison?  That would be very nice, but probably out of scope for now.